File: src/main/java/com/qsoftware/bubbles/registry/object/RegistryObject.java
package com.qsoftware.bubbles.registry.object;

import com.qsoftware.bubbles.common.IRegistryEntry;
import com.qsoftware.bubbles.common.RegistryEntry;
import com.qsoftware.bubbles.common.ResourceLocation;
import com.qsoftware.bubbles.registry.Registry;

import java.util.function.Supplier;

@SuppressWarnings({"rawtypes", "unchecked"})
public class RegistryObject<B extends IRegistryEntry> {
    private final Registry registry;
    private final Supplier<B> supplier;
    private final ResourceLocation resourceLocation;

    public <T extends B> RegistryObject(Registry registry, Supplier<B> supplier, ResourceLocation resourceLocation) {
        this.registry = registry;
        this.supplier = supplier;
        this.resourceLocation = resourceLocation;
    }

    public void register() {
        registry.register(resourceLocation, (RegistryEntry) ((Supplier) supplier).get());
    }

    @SuppressWarnings("unchecked")
    public B get() {
        return (B) registry.get(resourceLocation);
    }
}

File: src/main/kotlin/qtech/bubbles/registry/object/RegistryObject.kt
package qtech.bubbles.registry.`object`

import qtech.bubbles.common.IRegistryEntry
import qtech.bubbles.common.ResourceLocation
import qtech.bubbles.registry.Registry
import java.util.function.Supplier

class RegistryObject<B : IRegistryEntry>(private val registry: Registry<in B>, private val supplier: Supplier<B>, private val resourceLocation: ResourceLocation) {
    fun register() {
        registry.register(resourceLocation, supplier.get())
    }

    @Suppress("UNCHECKED_CAST")
    fun get(): B {
        return registry[resourceLocation] as B
    }
}